### ISSUE TITLE
Debug TLS issue?

### ISSUE THREAD

**Issue by sb10** [go-mangos/mangos#312]
I've been using rep/req over TCP just fine, and am trying to switch to tls+tcp, but the req (client) end can't Recv.

I do something like (err handling skipped for brevity, but no errors generated except at the end):

    sock, _ := req.NewSocket()
    sock.SetOption(mangos.OptionMaxRecvSize, 0)
    sock.SetOption(mangos.OptionRecvDeadline, timeout)
    sock.AddTransport(tlstcp.NewTransport())
    tlsConfig := &tls.Config{}
    caCert, _ := ioutil.ReadFile(caFile)
    certPool := x509.NewCertPool()
    certPool.AppendCertsFromPEM(caCert)
    tlsConfig.RootCAs = certPool
    // tlsConfig.InsecureSkipVerify = true
    dialOpts := make(map[string]interface{})
    dialOpts[mangos.OptionTLSConfig] = tlsConfig
    sock.DialOptions("tls+tcp://localhost:11301", dialOpts)
    [...]
    resp, err := sock.Recv()

With InsecureSkipVerify line uncommented it works, but with it commented out, the Recv() err is a generic `receive time out`. How do I get more information on what is really going wrong?

The rep (server) end seems fine. I do something like (again, err handling excluded):

    sock, _ := rep.NewSocket()
    sock.SetOption(mangos.OptionMaxRecvSize, 0)
    sock.SetOption(mangos.OptionRaw, true)
    sock.SetOption(mangos.OptionRecvDeadline, ServerInterruptTime)
    sock.AddTransport(tlstcp.NewTransport())
    cer, _ := tls.LoadX509KeyPair(serverPemFile, serverKeyFile)
    tlsConfig := &tls.Config{Certificates: []tls.Certificate{cer}}
    listenOpts := make(map[string]interface{})
    listenOpts[mangos.OptionTLSConfig] = tlsConfig
    sock.ListenOptions("tls+tcp://0.0.0.0:"+config.Port, listenOpts)

 I create the certs like this:

    func GenerateCerts(caFile, serverPemFile, serverKeyFile string) error {
    	// key for root CA
    	rootKey, err := rsa.GenerateKey(rand.Reader, 2048)
    	if err != nil {
    		return err
    	}

    	// cert for root CA, self-signed
    	rootCertTmpl, err := certTemplate()
    	if err != nil {
    		return err
    	}
    	rootCertTmpl.IsCA = true
    	rootCertTmpl.KeyUsage = x509.KeyUsageCertSign | x509.KeyUsageDigitalSignature
    	rootCertTmpl.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth, x509.ExtKeyUsageClientAuth}
    	rootCert, err := createCert(rootCertTmpl, rootCertTmpl, &rootKey.PublicKey, rootKey, caFile)
    	if err != nil {
    		return err
    	}

    	// key for server
    	serverKey, err := rsa.GenerateKey(rand.Reader, 2048)
    	if err != nil {
    		return err
    	}

    	// cert for server, signed by root CA
    	servCertTmpl, err := certTemplate()
    	if err != nil {
    		return err
    	}
    	servCertTmpl.KeyUsage = x509.KeyUsageDigitalSignature
    	servCertTmpl.ExtKeyUsage = []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth}
    	_, err = createCert(servCertTmpl, rootCert, &serverKey.PublicKey, rootKey, serverPemFile)
    	if err != nil {
    		return err
    	}

    	// store the server's key
    	keyOut, err := os.OpenFile(serverKeyFile, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
    	if err != nil {
    		return err
    	}
    	err = pem.Encode(keyOut, &pem.Block{
    		Type: "RSA PRIVATE KEY", Bytes: x509.MarshalPKCS1PrivateKey(serverKey),
    	})
    	if err != nil {
    		return err
    	}
    	return keyOut.Close()
    }

    // certTemplate creates a certificate template with a random serial number,
    // valid from now until validFor.
    func certTemplate() (*x509.Certificate, error) {
    	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
    	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
    	if err != nil {
    		return nil, errors.New("failed to generate serial number: " + err.Error())
    	}

    	template := x509.Certificate{
    		SerialNumber:          serialNumber,
    		Subject:               pkix.Name{Organization: []string{"my org"}},
    		SignatureAlgorithm:    x509.SHA256WithRSA,
    		NotBefore:             time.Now(),
    		NotAfter:              time.Now().Add(validFor),
    		IPAddresses:           []net.IP{net.ParseIP("0.0.0.0"), net.ParseIP("127.0.0.1")},
    		DNSNames:              []string{"localhost"},
    		BasicConstraintsValid: true,
    	}
    	return &template, nil
    }

    // createCert creates a certificate given a template, signing it against its
    // parent, and saving the cert in PEM format to certPath.
    func createCert(template, parentCert *x509.Certificate, publicKey interface{}, parentPrivateKey interface{}, certPath string) (*x509.Certificate, error) {
    	certDER, err := x509.CreateCertificate(rand.Reader, template, parentCert, publicKey, parentPrivateKey)
    	if err != nil {
    		return nil, err
    	}

    	// parse the resulting certificate so we can use it again
    	cert, err := x509.ParseCertificate(certDER)
    	if err != nil {
    		return nil, err
    	}

    	// save in PEM format
    	b := pem.Block{Type: "CERTIFICATE", Bytes: certDER}
    	certOut, err := os.Create(certPath)
    	if err != nil {
    		err = fmt.Errorf("creation of certificate file [%s] failed: %s", certPath, err)
    		return nil, err
    	}
    	err = pem.Encode(certOut, &b)
    	if err != nil {
    		return nil, err
    	}
    	err = certOut.Close()
    	return cert, err
    }

Any thoughts on how I can get it to work?

**Comment by gdamore** [comment]
I'm not sure if validation of IP addresses is supported -- you probably cannot validate 0.0.0.0 because that makes no sense (and would be otherwise insecure as 0.0.0.0 is a wild card meaning any address).

Server names should be established, and you should connect via name, I think.  The name can itself be an IP address if I recall properly.

Also, I suspect that your client is trying to validate not the IP address, but the name "localhost".

**Comment by sb10** [comment]
Sure, the client is given "localhost", and the cert is valid for "localhost", so it should work. I use the same server cert and key for a REST API on the same server, and `curl --cacert <caFile path> https://localhost:[...]` works fine.

**Comment by gdamore** [comment]
Is your certificate valid for TLS 1.2?  (I don't see any reason it wouldn't be...)

The other thing is that HTTPS may validate certificates differently -- I'm not sure the DNSNames field is honored when using TLS directly the way mangos does.  I don't see anywhere that you have set the Common Name (CN) in the certificate, nor the subjectAltName.   At least one of these should be set and include localhost or whatever name you are using to connect to.

I don't know what software considers DNSNames in the certificates -- this is the first time I've seen that used.

**Comment by gdamore** [comment]
Hmm... I'm also not seeing any sign that you're setting the ServerName field of the TLS configuration object.  Maybe that's the main problem?

**Comment by gdamore** [comment]
Yeah, I'm pretty sure it's missing ServerName that is the issue.  See also #252 -- you're not the first person to have this problem.

I'd really like to find a way to make these failures happen more synchronously with a more useful error to facilitate debug.

**Comment by sb10** [comment]
Thanks, I replaced `tlsConfig.InsecureSkipVerify = true` with `tlsConfig.ServerName = "localhost"` for the client tls.Config and it works.

It's worth noting that mongos' tls tests use generated certs: https://github.com/go-mangos/mangos/blob/c3916ba23b3aa721fd13b4dec916b2d15f4a2520/test/certs.go but there is no mention of ServerName in the codebase. I think the clients are simply configured with InsecureSkipVerify. Perhaps some tests of using TLS in "secure mode" are warranted.

I agree that a useful error message would be good, but in the meantime, perhaps added docs with a clear example of successfully working with a generated cert without InsecureSkipVerify would solve the issue.
